### DISCUSSION TITLE
Past and future events

### DISCUSSION THREAD

**Discussion by butkandrius** [JasperFx/marten#2442] [Q&A]
Hi, I am wondering if I can use Marten Event to store according this use case:
Basically I need to ability to add events in time point that is in past or future.
So when a field is updated I need to tell when this change is going to be established. As I see now  `event_dt` table has `timestamp` column that tells when operation was triggered. I can modify this field to suit this scenario by
inheriting from `DocumentSessionListenerBase `and overrding `BeforeSaveChangesAsync` so it could change  `timestamp`  data to the value that in Header (I would add new date value bofore that). But still I need a column that tells when this change actually was made and I don't see if it is possible to extend this table in Marten.
It looks like the scenario for handling future events it's not the right fit and I have to make customized event sourcing. Also I see another issue - read model can be invalid because there will be pending events in the future that means I need to update read model every day using job to apply future events to read model .

**Comment by krmapowel** [comment]
I've done some research for what are the solutions for handling events in time, be it in the past or in the future, and how is it done across different solutions.
I'd say that this problem is not Marten specific, it looks like an inherit "limitation" (or rather base assumption) of Event Store in general - Event is registered always at the end of the stream of all events.

In Event Store context, I think Greg Young prelection about consistency give some insights to that. In Postgres space I've seen some research done about Bi-Temporal tables.

For Events, you need to better describe what the event is. Sometimes what you're modeling is not Event that just happened, but additionally an Event of "Noticed that Event happened in the past" or "Assume that Event will happen in the future". In this situation you have Event Added Date (event_dt in this case) and Event In Effect Date. I don't believe that any Event Store should simplify this concept to just override Event Added Date to be used as Event In Effect Date.
If you add your event with another field Effect Date, then you will be able to solve this problem in projection and correctly put it in time based on current knowledge of timeline. It would allow you to query this projection document by comparing to DateTime.Now(), to see what is your current knowledge about things that already happened.

**Comment by agross** [comment]
This discussion might shed some light on time on domain modeling: https://youtu.be/xABtweT7Jzk?t=1274
